data/tables/aggregate.js

```diff
@@ -0,0 +1,206 @@
+ * @typedef {import('../types').CarItemAggregate} CarItem
+ */
+
+/** @type {Record<string, AggregateStat>} */
+export const AGGREGATE_STAT = {
+  ingesting: 'INGESTING',
+  ready: 'READY',
+  dealPending: 'DEAL_PENDING',
+  dealProcessed: 'DEAL_PROCESSED'
+}
+
+const DEFAULT_MAX_SIZE = 127*(1<<28)
+const DEFAULT_MIN_SIZE = 1+127*(1<<27)
+
+
+/**
+ * Abstraction layer to handle operations for aggregates to send to Spade.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {AggregateOpts} [options]
+ * @returns {import('../types').AggregateTable}
+ */
+export function createAggregateTable (region, tableName, options = {}) {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint
+  })
+  const minSize = options.minSize || DEFAULT_MIN_SIZE
+  const maxSize = options.maxSize || DEFAULT_MAX_SIZE
+
+  return {
+    /**
+     * Add new CARs to the given aggregate if still with enough space.
+     *
+     * @param {string} aggregateId
+     * @param {CarItem[]} cars 
+     */
+    add: async (aggregateId, cars) => {
+      const links = cars.map(car => car.link)
+      const updateAccumSize = cars.reduce((acc, car) => acc + car.size, 0)
+      const maxSizeBeforeUpdate = maxSize - updateAccumSize
+
+      if (maxSizeBeforeUpdate < 0) {
+        throw new Error('Given cars do not fit inside the given aggregate')
+      }
+
+      const insertedAt = new Date().toISOString()
+
+      const updateItemcommand = new UpdateItemCommand({
+        TableName: tableName,
+        Key: marshall({
+          aggregateId
+        }),
+        ExpressionAttributeValues: {
+          ':insertedAt': { S: insertedAt },
+          ':updatedAt': { S: insertedAt },
+          ':ingestingStat': { S: AGGREGATE_STAT.ingesting },
+          ':initialSize': { N: `${0}` },
+          ':updateAccumSize': { N: `${updateAccumSize}` },
+          ':maxSizeBeforeUpdate': { N: `${maxSizeBeforeUpdate}` },
+          ':links' :{ SS: links } // SS is "String Set"
+        },
+        // Condition expression runs before update. Guarantees that this operation only suceeds:
+        // - when stat is ingesting
+        // - when there is still enough space in the aggregate for this batch size
+        ConditionExpression: `
+        (
+          attribute_not_exists(stat) OR stat = :ingestingStat
+        )
+        AND
+        (
+          attribute_not_exists(size) OR size <= :maxSizeBeforeUpdate
+        )
+        `,
+        // Update row table with:
+        // - insertedAt if row does not exist already
+        // - updatedAt updated with current timestamp
+        // - set state as ingesting if row does not exist already
+        // - increment size
+        UpdateExpression: `
+        SET insertedAt = if_not_exists(insertedAt, :insertedAt),
+          updatedAt = :updatedAt,
+          stat = if_not_exists(stat, :ingestingStat),
+          size = if_not_exists(size, :initialSize) + :updateAccumSize
+        ADD cars :links
```

Comment:
So, what we know:
- Maximum item size is 400KB https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ServiceQuotas.html#limits-items
- SET has no max limit of entries https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ServiceQuotas.html#limits-attributes

With these we can do some maths:
- An array of strings with 100 CIDs is around 6400 bytes, so we can consider 1k CIDs to be around 64KB and 6250 CIDs to be around 400KB
- If we consider these 6250 CIDs to have an average size of 3MB, it already is more than enough for a deal.

I could not find metrics on average upload size at this moment. I think these numbers should be good for us, given if we get in an edge case it will retry with multiple batches (while also bisecting them) and it will end up go though at some point.

If we see this to happen (we likely will see a slow down on throughput), we can consider writing this into S3. I would like to avoid this given we will lose transaction guarantees by bring another storage unit in this call.

Reply:
For the record, we decided this was enough of a worry that we'd pull this column out into a table mapping aggregate ID to CAR CID and use [TransactWriteItems](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactWriteItems.html) to write to both the mapping table and the aggregate table in a single transaction.